src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BedSearchService.kt

```diff
@@ -0,0 +1,134 @@
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PostcodeDistrictRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRole
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.validated
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.ApprovedPremisesBedSearchResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.BedSearchRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.TemporaryAccommodationBedSearchResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.AuthorisableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.ValidatableActionResult
+import java.time.LocalDate
+
+@Service
+class BedSearchService(
+  private val bedSearchRepository: BedSearchRepository,
+  private val postcodeDistrictRepository: PostcodeDistrictRepository,
+  private val characteristicService: CharacteristicService,
+) {
+  fun findApprovedPremisesBeds(
+    user: UserEntity,
+    postcodeDistrictOutcode: String,
+    maxDistanceMiles: Int,
+    startDate: LocalDate,
+    durationInDays: Int,
+    requiredPremisesCharacteristics: List<String>,
+    requiredRoomCharacteristics: List<String>,
+  ): AuthorisableActionResult<ValidatableActionResult<List<ApprovedPremisesBedSearchResult>>> {
+    if (!user.hasRole(UserRole.MATCHER)) {
+      return AuthorisableActionResult.Unauthorised()
+    }
+
+    return AuthorisableActionResult.Success(
+      validated {
+        val premisesCharacteristicErrors = mutableListOf<String>()
+        val premisesCharacteristics = requiredPremisesCharacteristics.mapNotNull {
+          val characteristic = characteristicService.getCharacteristicByPropertyName(it)
+
+          if (characteristic == null) {
+            premisesCharacteristicErrors += "$it doesNotExist"
+            return@mapNotNull null
+          }
+
+          if (characteristic.serviceScope != ServiceName.approvedPremises.value || characteristic.modelScope != "premises") {
+            premisesCharacteristicErrors += "$it scopeInvalid"
+            return@mapNotNull null
+          }
+
+          characteristic
+        }
+
+        if (premisesCharacteristicErrors.any()) {
+          "$.requiredPremisesCharacteristics" hasValidationError premisesCharacteristicErrors.joinToString(", ")
+        }
+
+        val roomCharacteristicErrors = mutableListOf<String>()
+        val roomCharacteristics = requiredRoomCharacteristics.mapNotNull {
+          val characteristic = characteristicService.getCharacteristicByPropertyName(it)
+
+          if (characteristic == null) {
+            roomCharacteristicErrors += "$it doesNotExist"
+            return@mapNotNull null
+          }
+
+          if (characteristic.serviceScope != ServiceName.approvedPremises.value || characteristic.modelScope != "room") {
```

Comment:
This won't correctly handle characteristics that have either a service or model scope of `"*"`.